### PR TITLE
Combine `ray_julia_jll` deps/gen directories

### PR DESCRIPTION
**Pull request by omus**
While working on #158 I decided this was something we should also tackle. The rational behind this is that the `deps` directory in Julia is usually reserved for runtime dependencies of the package that are automatically built during package installation. As we're moving towards using artifacts only developers of Ray.jl will be required to build the "ray_julia" shared library and utilize `~/.julia/artifacts/Overrides.toml`. Due to this it seems best to combine the current artifact building directory `gen` and `deps` together into a new directory: `build`.

I also considered using the directory name `artifacts` but thought it could confuse users into thinking that the artifacts were embedded into the package.

### CODE REVIEW

**Comment by omus** [comment]
I was going to include this change in #158 but decided against it as both of these PRs include a non-trivial amount of changes.

File: ray_julia_jll/build/build.jl
Comment (kleinschmidt) [inline ray_julia_jll/build/build.jl:1]: I wonder if we want to keep this one as `build_jll` given that there are other `build_*` in here?

File: ray_julia_jll/build/build_tarballs.jl [hunk 1]
```diff
@@ -19,13 +19,13 @@ if abspath(PROGRAM_FILE) == @__FILE__
     isdir(TARBALL_DIR) || mkdir(TARBALL_DIR)
 
     @info "Building ray_julia_jll..."
-    run(`$(Base.julia_cmd()) --project=$JLL_PATH $JLL_PATH/deps/build_jll.jl `)
+    include("build.jl")
```
Comment (kleinschmidt) [inline ray_julia_jll/build/build_tarballs.jl:22]: are we sure that we don't need to fork a new julia process to run this build step?  why were we doing that in the first place?

File: Dockerfile [hunk 2]
```diff
@@ -178,22 +178,22 @@ RUN --mount=type=cache,sharing=locked,target=${BAZEL_CACHE},uid=${UID},gid=${GID
...
     #
     # Build using the final Ray.jl destination
-    ln -s ${RAY_REPO} ${RAY_JLL_REPO}/deps/ray && \
+    ln -s ${RAY_REPO} ${RAY_JLL_REPO}/build/ray && \
     rm -rf ${RAY_JLL_PROJECT} && \
     ln -s ${RAY_JLL_REPO} ${RAY_JLL_PROJECT} && \
     #
     # Build ray_julia_jll
-    julia --project=${RAY_JLL_PROJECT} -e 'using Pkg; Pkg.build(verbose=true); Pkg.precompile(strict=true)' && \
+    julia --project=${RAY_JLL_PROJECT} -e 'using Pkg; Pkg.build(); Pkg.precompile(strict=true)' && \
```
Comment (kleinschmidt) [inline Dockerfile:186]: can this still be done w/ `Pkg.build` or do we need to actually `include` the build.jl file?

File: ray_julia_jll/.gitignore [hunk 3]
```diff
@@ -1,9 +1,7 @@
...
-deps/build.log
-deps/WORKSPACE.bazel
-deps/ray
-deps/bazel-*
-deps/override
 /.dir-locals.el
-gen/tarballs
+build/WORKSPACE.bazel
+build/ray
+buid/bazel-*
```
Comment (kleinschmidt) [inline ray_julia_jll/.gitignore:6]:
```suggestion
build/bazel-*
```

File: ray_julia_jll/build/README.md
Comment (kleinschmidt) [inline ray_julia_jll/build/README.md:1]: this needs to be updated as well with the changes to the filenames (make.jl -> build_tarballs.jl) for instance

**Review by kleinschmidt — APPROVED** [review]
a few Qs and small things but otherwise looks good

File: ray_julia_jll/build/build.jl
Comment (omus) [reply ray_julia_jll/build/build.jl:1]: I think `build_library.jl` would be the name to use if we did this

File: ray_julia_jll/build/build_tarballs.jl (hunk 1 above)
Comment (omus) [reply ray_julia_jll/build/build_tarballs.jl:22]: The reason we previously did this was due to the `build_jll.jl` and the the `make.jl` having separate projects with different dependencies.

File: Dockerfile (hunk 2 above)
Comment (omus) [reply Dockerfile:186]: The `Dockerfile` needs to include the `build.jl` now and should have included the `build_jll.jl` when that was added. Having #155 would help catch these things in the future. I'll be making another PR which fixes the Dockerfile

Comment (omus) [reply Dockerfile:186]: Opened #160 for this. As the Dockerfile is currently broken on `main` I think I'll get this merged so we can keep things moving.

**Pull request merged**